TF-M: Add option to enable SFN model

joerchan opened the pull request:
Add option to enable the SFN model when building TF-M.
The SFN model will eventually replace the Library model.

Change the default model to be IPC, which follows the default
configuration of TF-M.

Allow building for nRF non-secure and TF-M enabled without having the
platform partition enabled.

joerchan committed:
tfm: Add option to enable the SFN model

Add option to enable the SFN model when building TF-M.
The SFN model will eventually replace the Library model.

Change the default model to be IPC, which follows the default
configuration of TF-M.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
soc: nordic_nrf: Allow building with TF-M without platform partition

Allow building for nRF non-secure and TF-M enabled without having the
platform partition enabled.
In this configuration the soc_secure functions have no valid
implementation, leave it as a compilation error if the application
includes the functions.

Add missing dependency on soc_secure functions being available for
SOC_HFXO_CAP_INTERNAL configuration.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

microbuilder commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -276,14 +276,54 @@ config TFM_MCUBOOT_DATA_SHARING
-	  TF-M Profile option is supplied, do not allow manual setting of
-	  the IPC mode, as it is determined by the profile setting.
+	  the PSA API using the SFN Model as the SPM backend.
+	  The SFN model support the SFN Partition model, and the isolation
+	  levels 1.
```
Super nit, but since there is a CI failure we may as well fix this at the same time as any update:

```suggestion
	  The SFN model supports the SFN Partition model, and isolation
	  level 1.
```

joerchan replied:
Spelling is Kconfig help text is not nit ;)

microbuilder reviewed (changes requested):
Tiny nit, sorry, but requesting changes since CI failing anyway.

microbuilder commented:
@theotherjimmy Do you have any bandwidth to have a look at that failure on the lpc55s69, since you have some experience with that part, and I'm not sure @joerchan has one available?

joerchan commented:
@microbuilder @theotherjimmy I suspect the lpc55 issue is related to TFM_IPC being default now, and this has an issue for the board. Maybe we could check if this is an issue on the main branch as well.

We could put this board as default Library model for the time and open an issue to resolve it later.

joerchan commented:
@theotherjimmy It is a kconfig / cmake issue, so you don't have too look into this problem. I can fix this myself.

joerchan commented:
CI Issue resolved. Conditional prompt on a choice symbol meant that none of the options was selected, which is not a valid.

SebastianBoe commented on soc/arm/nordic_nrf/common/soc_secure.c:
```diff
@@ -9,23 +9,28 @@
 	__ASSERT(err == TFM_PLATFORM_ERR_SUCCESS, "TFM platform error (%d)", err);
 	__ASSERT(result == 0, "GPIO service error (%d)", result);
+#endif
 }
```
Some static analyzers / compilers complain about unused variables.

Shall we add some ARG_UNUSED() macro's here ?

SebastianBoe replied:
N/A if we use a dedicated file. See other feedback.

joerchan replied:
Addressed by excluding these functions from the build
There should be proper Kconfig dependencies to produce an error at config time if attempting to setup an invalid configuration.

SebastianBoe commented on soc/arm/nordic_nrf/common/soc_secure.c:
```diff
@@ -44,4 +49,7 @@ int soc_secure_mem_read(void *dst, void *src, size_t len)
 	}
+#else
+	return -ENOTSUP;
+#endif
```
There is no code-reuse in this file between CONFIG_TFM_PARTITION_PLATFORM and not CONFIG_TFM_PARTITION_PLATFORM.

I suggest we add a soc_secure_dummy.c file to simplify the code of the common case.

joerchan replied:
Change reverted.

SebastianBoe commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -276,14 +276,53 @@ config TFM_MCUBOOT_DATA_SHARING
+	bool "Library Model"
+	help
+	  When enabled this option signifies that the TF-M build use secure
+	  library model instead of PSA API.
```
```suggestion
	  library model instead of the IPC Model.
```

The PSA API is supported in the library model.

SebastianBoe replied:
Strange, but insignificant I guess since this is going to be deleted in a few months time.

SebastianBoe commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -276,14 +276,53 @@ config TFM_MCUBOOT_DATA_SHARING
-	default y if (TFM_PROFILE_TYPE_MEDIUM || TFM_PROFILE_TYPE_LARGE)
+	bool "IPC Model"
+	help
+	  When enabled, this option signifies that the TF-M build supports
+	  the PSA API using the IPC Model as the SPM backend.
```
```suggestion
	  Use the IPC Model for the PSA API.
```

It's bad practice to start Kconfig help texts with "when enabled" etc.

SebastianBoe replied:
same for SFN

joerchan replied:
Reworded suggestion.

SebastianBoe commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -276,14 +276,53 @@ config TFM_MCUBOOT_DATA_SHARING
+	  v1.1.
+
+config TFM_SFN
+	bool "SFN model"
```
depends on ISOLATION_LEVEL_1 (don't know the name)

?

SebastianBoe replied:
ah, right you are. Promptless for non-IPC I see.

```
config TFM_ISOLATION_LEVEL
	int "Isolation level setting." if (TFM_PROFILE_TYPE_NOT_SET && TFM_IPC)
	range 1 3
	depends on BUILD_WITH_TFM
	default 1 if TFM_PROFILE_TYPE_SMALL || !TFM_IPC

```

SebastianBoe commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -276,14 +276,53 @@ config TFM_MCUBOOT_DATA_SHARING
+	  the PSA API using the SFN Model as the SPM backend.
+	  The SFN model supports the SFN Partition model, and isolation level 1.
+
+	  In this model the Secure Partition is made up of a collection of
```
```suggestion
	  In this model each Secure Partition is made up of a collection of
```

SebastianBoe commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -276,14 +276,53 @@ config TFM_MCUBOOT_DATA_SHARING
+	  The IPC model supports the IPC and SFN Partition models, and
+	  isolation levels 1, 2 and 3.
+
+	  In this model the Secure Partition processes signals in any order,
```
```suggestion
	  In this model each Secure Partition processes signals in any order,
```

SebastianBoe commented on modules/trusted-firmware-m/Kconfig.tfm.partitions:
```diff
@@ -52,6 +56,7 @@ config TFM_PARTITION_INITIAL_ATTESTATION
 config TFM_PARTITION_PLATFORM
 	bool "Secure partition 'Platform'"
 	default y
+	depends on !TFM_SFN # Using PSA Framework version 1.0
```
```suggestion
	depends on !TFM_SFN # Currently using PSA Framework version 1.0
```

to indicate this is expected to be fixed soon

SebastianBoe reviewed (changes requested):
changes requested

joerchan committed:
soc: nrf5340: Produce a Kconfig error when GPIO cannot be forwarded

Make Kconfig produce an error when GPIOs cannot be forwarded in the
current configuration instead of silently excluding the forwarding.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

SebastianBoe reviewed (dismissed):
Very nicely configured.

anangl commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -169,7 +163,7 @@ static int nordicsemi_nrf53_init(const struct device *arg)
 #endif
 
-#if defined(NRF_GPIO_FORWARDER_FOR_NRF5340_CPUAPP_ENABLED)
+#if defined(CONFIG_BOARD_NRF_GPIO_FORWARDER_FOR_NRF5340)
```
The `BOARD_NRF_GPIO_FORWARDER_FOR_NRF5340` option is defined as available only for `nrf5340dk_nrf5340_cpuapp*` targets, but also others, like `thingy53_nrf5340_cpuapp*` and `bl5340_dvk_cpuapp*`, need to use the code wrapped in this `#if`.

joerchan replied:
Thanks. I moved the Kconfig to SOC common, since it affects nrf53/soc.c, and then all boards need to select it on their BOARD_ENABLE_CPUNET.
